[site-isolation] Fix document-leak-test.js to work with frames in different processes

csaavedra opened the pull request:
#### 2289856d473aaff9720242175fef82acc7036a24
<pre>
[site-isolation] Fix document-leak-test.js to work with frames in different processes
<a href="https://bugs.webkit.org/show_bug.cgi?id=303237">https://bugs.webkit.org/show_bug.cgi?id=303237</a>

Reviewed by Ryosuke Niwa.

This adopts the fixes to http/tests/resources/document-leak-test.js from
r302683@main to the version in LayoutTests/resources, fixing with this
several tests that are failing when site isolation is enabled.

One important change from the version in http/tests/resources/ is that here
we run a proportional number of checks depending on how many frames
were loaded, since I noticed that a fixed number of checks are not always
going to be enough for GC to be triggered in time and are likely to cause
flakiness.

Also updated lazy-image-loading-document-leak.html to use this helper
instead of having its own implementation.

* LayoutTests/fast/dom/lazy-image-loading-document-leak-expected.txt:
* LayoutTests/fast/dom/lazy-image-loading-document-leak.html:
* LayoutTests/fast/dom/resources/lazy-image-loading-document-leak-frame.html: Renamed from LayoutTests/fast/dom/resources/lazy-image-loading-document-leak-popup.html.
* LayoutTests/platform/glib/media/media-source/remoteplayback-availability-callback-does-not-leak-expected.txt: Removed.
* LayoutTests/platform/ios-site-isolation/TestExpectations:
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* LayoutTests/resources/document-leak-test.js:
(createFrames):
(iframeLeaked): Deleted.
(iframeSentMessage):
(runDocumentLeakTest): Deleted.

Canonical link: <a href="https://commits.webkit.org/304240@main">https://commits.webkit.org/304240@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/89ed3ad7162cfdd2fb1371cb07f302e71a379491

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134969 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7386 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46219 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142478 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86818 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8007 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7232 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103131 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/70388 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137915 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5673 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120976 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83982 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5492 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3104 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3075 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39132 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145177 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7062 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39707 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111509 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7115 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5926 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111868 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28389 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5327 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117259 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60987 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7111 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35431 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6884 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7118 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6991 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on LayoutTests/resources/document-leak-test.js:
```diff
@@ -49,6 +49,8 @@ function iframeLeaked()
     }
 }
 
+var done = false;
```
I don't think I understand how you got to this.  Are you running tests with site isolation on?  How, of all the hundreds of things that are broken, did you decide to fix this one first?  How do you know this fixes them?  I would expect that not to fix anything because the "done" variable is scoped to the global object, which has no concept of what process it is in

csaavedra replied:
The way I ended up fixing this is that I was looking at the test results with site isolation on (in build.webkit.org) to pick something to fix. [One of the tests failing](https://build.webkit.org/results/Apple-Tahoe-Release-WK2-Site-Isolation-Tree-Tests/303819@main%20(629)/fast/css/css-anchor-position/anchor-position-does-not-leak-pretty-diff.html) there is failing because that logging from that function below is getting called multiple times and the logging is repeated, so I started checking why, and it ended up being this. I'm not claiming this is particularly important vs the other things that are broken, it just happened to be the first one I investigated. 🤷🏼 Turned out that this affects all tests that are using this document-leak-test.js (all failures for tests named `*does-not-leak.html` have the same issue, see [here](https://build.webkit.org/results/Apple-Tahoe-Release-WK2-Site-Isolation-Tree-Tests/303819@main%20(629)/results.html).

I might not have explained properly why this fix works; this code is running in the main frame's scope only (tests using `runDocumentLeakTest(options)` run this in the main frame, usually, and it's this helper the one creating many frames). Those frames created are loading some URI defined by the test. My understanding is that without site isolation `finishJSTest()` will finish the test after logging once after checking that the document doesn't leak; but with site isolation and subframes running in different processes, more messages from different frames before the test is finished, and so the logging code runs many times. With that (admittedly, ugly) global variable in the main frame global object, once we have logged for the first received message, we stop logging the ones that might later from frames in other processes.

And I know this fixes it because I ran those tests locally, with site isolation enabled, and they pass.

csaavedra committed:
[site-isolation] Fix document-leak-test.js to work with frames in different processes
https://bugs.webkit.org/show_bug.cgi?id=303237

Reviewed by Ryosuke Niwa.

This adopts the fixes to http/tests/resources/document-leak-test.js from
r302683@main to the version in LayoutTests/resources, fixing with this
several tests that are failing when site isolation is enabled.

One important change from the version in http/tests/resources/ is that here
we run a proportional number of checks depending on how many frames
were loaded, since I noticed that a fixed number of checks are not always
going to be enough for GC to be triggered in time and are likely to cause
flakiness.

Also updated lazy-image-loading-document-leak.html to use this helper
instead of having its own implementation.

* LayoutTests/fast/dom/lazy-image-loading-document-leak-expected.txt:
* LayoutTests/fast/dom/lazy-image-loading-document-leak.html:
* LayoutTests/fast/dom/resources/lazy-image-loading-document-leak-frame.html: Renamed from LayoutTests/fast/dom/resources/lazy-image-loading-document-leak-popup.html.
* LayoutTests/platform/glib/media/media-source/remoteplayback-availability-callback-does-not-leak-expected.txt: Removed.
* LayoutTests/platform/ios-site-isolation/TestExpectations:
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* LayoutTests/resources/document-leak-test.js:
(createFrames):
(iframeLeaked): Deleted.
(iframeSentMessage):
(runDocumentLeakTest): Deleted.

Canonical link: https://commits.webkit.org/304240@main
